src/qiboml/operations/expectation.py

```diff
@@ -128,11 +141,53 @@ def grad(upstream):
 
         return expval, grad
 
-    if differentiation_rule is not None:
-        return _expectation(params)
+    return _expectation(params)
 
-    elif nshots is None:
-        return _exact(observable, circuit, initial_state, exec_backend)
 
-    else:
-        return _with_shots(observable, circuit, initial_state, nshots, exec_backend)
+def _with_jax(
+    observable,
+    circuit,
+    initial_state,
+    nshots,
+    backend,
+    differentiation_rule,
+):
+    """
+    Compute expectation sample integrating the custom differentiation rule with
+    TensorFlow's automatic differentiation.
+    """
+    import jax  # pylint: disable=import-error
+
+    params = circuit.get_parameters()
+    exec_backend = construct_backend(backend)
+
+    @jax.custom_gradient
+    def _expectation(params):
+        params = jax.numpy.array(params)
+
+        def grad(params):
```

Comment:
How about jitting grad? Is it possible?

Reply:
It is, but in principle it could be done even later on: JAX transformations are composable, so you could do `grad(grad(jit(grad(...))))`.

Reply:
I `jitted` all the functions we are returning now, but I think we should build a Jax version of the Qibojit backend, with a more fine-grained jitting, to boost the performances.